### PR TITLE
Repeated Group Headers

### LINKED ISSUE #726: Handle repeated groups
**Is your feature request related to a problem? Please describe.**
Repeated question groups

**Describe the solution you'd like**
Potentiall we can add a "+" button so that the user can create new groups. Groups that are repeatable can also be deleted with a "-" button

**Describe alternatives you've considered**
NA

**Additional context**
NA

**Would you like to work on the issue?**
NA

### PR DESCRIPTION
**Pull request by kevinmost**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #726

**Description**
Add headers to repeated groups.

The header shows the index of the repeated group instance, plus a delete button that removes that entire instance.

**Type**
Choose one: Feature

**Screenshots (if applicable)**

[Screenshot](https://user-images.githubusercontent.com/3058312/235964476-cfaea73f-5d09-4ff0-a8d6-65deb5dc8494.png)

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by jingtang10** [comment]
@shelaghm please take a look at the screenshot!

we didn't really know what to do with the heading so used cardinal numbers 1,2,...

@kevinmost and i did wonder if there's a group header extension from the FHIR spec we can use... or maybe we can use a placeholder like Group 1, Group 2,...

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireResponses.kt [hunk 1]
```diff
@@ -119,3 +117,6 @@ private fun unpackRepeatedGroups(
     listOf(questionnaireResponseItem)
   }
 }
+
+internal val Questionnaire.QuestionnaireItemComponent.isRepeatedGroup: Boolean
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireResponses.kt:121]: need unit test coverage

Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireResponses.kt:121]: also this should be in a different class.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
great pr! love this ❤️. thanks @kevinmost

**Comment by shelaghm** [comment]
Looks amazing! @kevinmost
A few suggestions to make it even better:

1.  Add 'Group' in front of 1, 2. So it reads Group 1, Group 2, etc.
2. Move the `sample date question` text so that it is the placeholder text inside the text field and dropdown. 
3. Use the EntryFormat / hint text field for the date format. 

(See screenshot for details)
Repeat Group 12

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 2]
```diff
@@ -711,11 +712,23 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
       // For background, see https://build.fhir.org/questionnaireresponse.html#link.
       buildList {
           // Case 1
-          add(questionnaireResponseItem.item)
+          if (!questionnaireItem.isRepeatedGroup) {
```
[the remark this answers is not in the document]
Comment (kevinmost) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:805]: Aren't these cases already extensively covered? For example, `questionnaireHasNestedItem_ofTypeGroup_shouldNestItemWithinItem` hits this code-path.

**Comment by kevinmost** [comment]
Updated

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt [hunk 3]
```diff
@@ -154,6 +154,21 @@ data class QuestionnaireViewItem(
     )
   }
 
+  internal fun removeAnswerAt(index: Int) {
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt:164]: oh ok i see

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireEditAdapter.kt
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireEditAdapter.kt:1]: @kevinmost

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:805]: yeah i think you're right! i'm just sensitive to any change that doesn't trigger a change in the unit test.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:1]: are we missing a test for delete?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreViews.kt
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreViews.kt:1]: `MoreViewGroups.kt`?

**Review by jingtang10 — CHANGES_REQUESTED** [review]
thanks for this @kevinmost!

happy to approve once all comments have been addressed - don't think i have any more comments.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireAdapterItem.kt [hunk 4]
```diff
@@ -17,9 +17,20 @@
...
   data class Question(val item: QuestionnaireViewItem) : QuestionnaireAdapterItem
+
+  /** A row for a repeated group response instance's header. */
+  data class RepeatedGroupHeader(
+    /** The response index. This is 0-indexed, but should be 1-indexed when rendered in the UI. */
+    val index: Int,
+    /** Callback that is invoked when the user clicks the delete button. */
+    val onDeleteClicked: () -> Unit,
+    /** Responses nested under this header. */
+    val responses: List<QuestionnaireResponse.QuestionnaireResponseItemComponent>,
```
[the remark this answers is not in the document]
Comment (kevinmost) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireAdapterItem.kt:34]: We'll change `areItemsTheSame` to use the `questionnaireResponseItem`, and `areContentsTheSame` to use the `index`

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreViews.kt
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreViews.kt:1]: done.

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/contrib/views/PhoneNumberViewHolderFactoryInstrumentedTest.kt [hunk 5]
```diff
@@ -71,7 +71,8 @@ class PhoneNumberViewHolderFactoryInstrumentedTest {
           )
           .viewType,
       )
-    assertThat(viewHolderFromAdapter).isInstanceOf(viewHolder::class.java)
+    viewHolderFromAdapter as QuestionnaireEditAdapter.ViewHolder.QuestionHolder
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/contrib/views/PhoneNumberViewHolderFactoryInstrumentedTest.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireResponses.kt (hunk 1 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireResponses.kt]: done.

File: datacapture/src/main/res/layout/repeated_group_instance_header_view.xml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ limitations under the License.
+-->
+<LinearLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:layout_marginHorizontal="@dimen/item_margin_horizontal"
+    android:layout_marginVertical="@dimen/item_margin_vertical"
+    android:orientation="horizontal"
+    android:gravity="center_vertical"
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/res/layout/repeated_group_instance_header_view.xml:24]: The `android:gravity="center_vertical"` attribute is used to vertically center align the child views, such as the Group text (TextView) and Delete icon (ImageView), within the LinearLayout. According to Shelagh's mockup, the Group text is already aligned to the left by default.

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/contrib/views/PhoneNumberViewHolderFactoryInstrumentedTest.kt [hunk 7]
```diff
@@ -71,7 +71,8 @@ class PhoneNumberViewHolderFactoryInstrumentedTest {
...
+    assertThat(viewHolderFromAdapter.holder).isInstanceOf(viewHolder::class.java)
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/contrib/views/PhoneNumberViewHolderFactoryInstrumentedTest.kt]: done.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/test/QuestionnaireUiEspressoTest.kt [hunk 8]
```diff
@@ -604,6 +609,89 @@ class QuestionnaireUiEspressoTest {
...
+      val recyclerView = view as RecyclerView
+      var count = 0
+      for (i in 0 until recyclerView.adapter!!.itemCount) {
+        val holder = recyclerView.findViewHolderForAdapterPosition(i)
+        if (
+          holder?.itemView?.findViewById<View>(R.id.repeated_group_instance_header_title) != null
+        ) {
+          count++
+        }
+      }
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/test/QuestionnaireUiEspressoTest.kt]: done.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponentsTest.kt [hunk 9]
```diff
@@ -2478,6 +2478,30 @@ class MoreQuestionnaireItemComponentsTest {
...
+
+    assertThat(question.isRepeatedGroup).isTrue()
+  }
+
+  @Test
+  fun testQuestionnaireItemComponentIsNotRepeatedGroup() {
+    val question =
+      Questionnaire.QuestionnaireItemComponent().apply {
+        linkId = "1"
+        type = Questionnaire.QuestionnaireItemType.BOOLEAN
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponentsTest.kt]: done.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponentsTest.kt [hunk 10]
```diff
@@ -2478,6 +2478,30 @@ class MoreQuestionnaireItemComponentsTest {
...
+        linkId = "1"
+        type = Questionnaire.QuestionnaireItemType.GROUP
+        repeats = true
+      }
+
+    assertThat(question.isRepeatedGroup).isTrue()
+  }
+
+  @Test
+  fun testQuestionnaireItemComponentIsNotRepeatedGroup() {
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponentsTest.kt]: done.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponentsTest.kt [hunk 11]
```diff
@@ -2478,6 +2478,30 @@ class MoreQuestionnaireItemComponentsTest {
     assertThat(zipList.size).isEqualTo(3)
   }
 
+  @Test
+  fun testQuestionnaireItemComponentIsRepeatedGroup() {
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponentsTest.kt]: done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireEditAdapter.kt
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireEditAdapter.kt:1]: Works in the review mode.

[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireEditAdapter.kt:1]: Added `readonly' attribute implementation to repeated group.

[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireEditAdapter.kt:1]: In the existing implementation, a repeatedGroup only shows nested items if the "Add Item" button is clicked. If it is in Read Only mode, the "Add Item" button is not displayed, hence there is no way to show nested items. 
@jingtang10

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt]: @santosh-pingle please add a test case

**Comment by santosh-pingle** [comment]
https://github.com/google/android-fhir/pull/1994/files#diff-1f4ba456c28c5400e29ae89383faf8d1a8d9109c80299bc0efb1ed3cf11818b2R834

This change has side effects:

Nested items of repeated group items are not shown.
Answers that are captured in questionnaireResponse.json and passed as an argument to QuestionnaireFragment do not show nested items of repeated group items.

If the following fix is added, then the above-mentioned issues are not seen:
`add(questionnaireResponseItem.item)`

@jingtang10

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireAdapterItem.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinmost) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireAdapterItem.kt:34]: It sounds to me like we wanted to flip the `areItemsTheSame` and `areContentsTheSame` implementations?

i.e, `areItemsTheSame` should be `newItem is QuestionnaireAdapterItem.RepeatedGroupHeader && oldItem.responses == newItem.responses`

and `areContentsTheSame` should be `newItem is QuestionnaireAdapterItem.RepeatedGroupHeader && oldItem.index == newItem.index`

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt [hunk 12]
```diff
@@ -167,6 +167,21 @@ data class QuestionnaireViewItem(
     )
   }
 
+  internal suspend fun removeAnswerAt(index: Int) {
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt:170]: done.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 13]
```diff
@@ -4160,18 +4160,25 @@ class QuestionnaireViewModelTest {
 
       assertThat(
           viewModel.getQuestionnaireItemViewItemList().map {
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:4162]: local function was already renamed.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 14]
```diff
@@ -3565,12 +3565,12 @@ class QuestionnaireViewModelTest {
 
     fun repeatedGroupA() =
       viewModel.getQuestionnaireItemViewItemList().single {
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:4125]: this change was already updated in the codebase.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireAdapterItem.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireAdapterItem.kt:34]: Kevin had already updated this change, so resolving it.

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/test/QuestionnaireUiEspressoTest.kt [hunk 15]
```diff
@@ -604,6 +609,89 @@ class QuestionnaireUiEspressoTest {
...
+  fun test_repeated_group_is_deleted() {
+    buildFragmentFromQuestionnaire("/component_repeated_group.json")
+
+    onView(withId(R.id.questionnaire_edit_recycler_view))
+      .perform(
+        RecyclerViewActions.actionOnItemAtPosition<ViewHolder>(
+          0,
+          clickChildViewWithId(R.id.add_item),
+        ),
+      )
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/test/QuestionnaireUiEspressoTest.kt:650]: Acknowledged. 
I will check other way to add questionnaire response.
(If the fix is not ready by the time the PR gets merged, I will create a separate PR.)

[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/test/QuestionnaireUiEspressoTest.kt]: done, provided questionnaire response to the fragment.

**Review by jingtang10 — APPROVED** [review]

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/test/QuestionnaireUiEspressoTest.kt [hunk 16]
```diff
@@ -50,8 +56,7 @@ import com.google.common.truth.Truth.assertThat
 import java.math.BigDecimal
 import java.time.LocalDate
 import java.time.LocalDateTime
-import java.util.Calendar
-import java.util.Date
+import java.util.*
```
[the remark this answers is not in the document]
Comment (santosh-pingle) [reply datacapture/src/androidTest/java/com/google/android/fhir/datacapture/test/QuestionnaireUiEspressoTest.kt]: done.

**Pull request merged**